internal/bgp/frr/config.go

```diff
@@ -103,15 +117,39 @@ func writeConfig(config string, filename string) error {
 
 // reloadConfig requests that FRR reloads the configuration file. This is
 // called after updating the configuration.
-func reloadConfig() error {
-	// TODO
-	return nil
+var reloadConfig = func(sf *singleflight.Group) error {
+	_, err, _ := sf.Do("reload",
+		func() (interface{}, error) {
+			time.Sleep(2 * time.Second)
+			pidFile, found := os.LookupEnv("FRR_RELOADER_PID_FILE")
+			if found {
+				reloaderPidFileName = pidFile
+			}
+
+			pid, err := os.ReadFile(reloaderPidFileName)
+			if err != nil {
+				return nil, err
+			}
+
+			pidInt, err := strconv.Atoi(string(pid))
+			if err != nil {
+				return nil, err
+			}
+
+			// send HUP signal to FRR reloader
+			err = syscall.Kill(pidInt, syscall.SIGHUP)
+			if err != nil {
+				return nil, err
+			}
+			return nil, nil
+		})
+	return err
```

Comment:
this is interesting package so the trick here is the sleep 2s u wanted to prevent doing back2back reload for 2sec window ?

Reply:
Single flight squash all the calls that happen while one is in progress.
Adding the sleep will make the call last longer and will squash all the configuration requests into a single one (with the drawback of 2 seconds timeout). As a second thought, I can make the timeout even shorter since we care about squashing the calls caused by the same config change.

Reply:
does it squash or block till on flight call completes ?

Reply:
The point about blocking makes total sense and I did not think about it. This will be likely to be called many subsequent times from the single goroutine, so the effect will be of delaying instead of really squashing them.
Let me put a goroutine in the middle so it will sort out the desired effect.

Reply:
is there any possibility for this working async now that we could fire frr reload with hlaf baked config ? from what I have seen FRR will drop the entire config if one line is incorrect, just thinking loud here since we don't have any idea when the reload took effect debugging invalid config will be interesting

Reply:
Configuration updates happen in discrete steps, and each step is supposed to provide a valid config (not meaning that we won't have bugs). What I am saying is that we won't produce a half baked config, we'll produce a config with less stuff that the caller expects:

What could happen here, is that the timeout falls in the middle (i.e. we are looping here https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/bgp_controller.go#L172 and the threshold happens half way through https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/bgp_controller.go#L269 )

With that, we would have less addresses announced, but the configuration (bugs aside) will still be valid.

In any case, I am torn about this because as it is now it's just introducing a delay and not working asynchronously and I don't want this to hold the integration to get in. Will remove the commit and rework it in a separate PR, using goroutines.

Reply:
Removed the reloading optimization, will handle it in a separate PR